### PR TITLE
Fixing https://github.com/googlesamples/unity-jar-resolver/issues/48

### PR DESCRIPTION
**Pull request by yuriy-universal-ivanov**
Fix for https://github.com/googlesamples/unity-jar-resolver/issues/48

### CODE REVIEW

**Comment by yuriy-universal-ivanov** [comment]
I signed it!

image
Initially I used a different (non-google) primary email address with this GitHub account, but I've just switched the primary email account to the same as used when signing CLA.

File: source/JarResolverLib/src/Google.JarResolver/PlayServicesSupport.cs [hunk 1]
```diff
@@ -1038,7 +1036,10 @@ public void ClearDependencies()
                 KeyValuePair<Dependency, string> oldDepFilenamePair;
                 if (currentDepsByVersionlessKey.TryGetValue(dep.VersionlessKey,
                                                             out oldDepFilenamePair)) {
-                    if (oldDepFilenamePair.Key.BestVersion != dep.BestVersion &&
+                    string oldVersion = ExtractVersionFromFileName(
```
Comment (stewartmiles) [inline source/JarResolverLib/src/Google.JarResolver/PlayServicesSupport.cs]: Dependency.BestVersion should be the version string so in both cases this will be either a version string or an empty string see https://github.com/googlesamples/unity-jar-resolver/blob/master/source/JarResolverLib/src/Google.JarResolver/Dependency.cs#L158

Why did you need to change this to extract the version from the filename?  Are you seeing a file name sneak into this version field?

File: source/JarResolverLib/src/Google.JarResolver/PlayServicesSupport.cs [hunk 2]
```diff
@@ -502,12 +502,11 @@ public void ClearDependencies()
 
                         // Extract the version from the filename.
                         // dep.Artifact is the name of the package (prefix)
-                        // The regular expression extracts the version number from the filename
-                        // handling filenames like foo-1.2.3-alpha.
-                        match = System.Text.RegularExpressions.Regex.Match(
-                            filenameWithoutExtension.Substring(
-                                dep.Artifact.Length + 1), "^([0-9.]+)");
-                        if (match.Success)
+                        string artifactVersion = ExtractVersionFromFileName(
```
Comment (stewartmiles) [inline source/JarResolverLib/src/Google.JarResolver/PlayServicesSupport.cs]: Not sure why you need this.  The previous code attempted to extract the version from the filename and if the match succeeds then jumps into the block below and pulls the version from match.Groups[1].Value.  Did you see a problem in this code path or was this just a clean up?

File: source/JarResolverLib/src/Google.JarResolver/PlayServicesSupport.cs [hunk 3]
```diff
@@ -1038,7 +1036,10 @@ public void ClearDependencies()
                 KeyValuePair<Dependency, string> oldDepFilenamePair;
                 if (currentDepsByVersionlessKey.TryGetValue(dep.VersionlessKey,
                                                             out oldDepFilenamePair)) {
-                    if (oldDepFilenamePair.Key.BestVersion != dep.BestVersion &&
+                    string oldVersion = ExtractVersionFromFileName(
+                        oldDepFilenamePair.Key.BestVersion);
+                    string newVersion = ExtractVersionFromFileName(dep.BestVersion);
+                    if ((oldVersion == null || (newVersion != null && oldVersion != newVersion)) &&
                         (confirmer == null || confirmer(oldDepFilenamePair.Key, dep))) {
```
Comment (stewartmiles) [inline source/JarResolverLib/src/Google.JarResolver/PlayServicesSupport.cs]: The stack trace reported in bug #48 starts here and bubbles up through to Resolver1_1.AarPathToPackageName where - from your report - the filename was null and therefore caused a null reference exception.  Which suggests https://github.com/googlesamples/unity-jar-resolver/blob/master/source/PlayServicesResolver/src/ResolverVer1_1.cs#L723 is returning a null path (as artifactPath) instead.

**Comment by yuriy-universal-ivanov** [comment]
See https://github.com/googlesamples/unity-jar-resolver/issues/48#issuecomment-299999696

**Comment by stewartmiles** [comment]
Ah ok, I see now.  Any chance you could update the commit message to reflect the comment you added here https://github.com/googlesamples/unity-jar-resolver/issues/48#issuecomment-299999696 ?

**Comment by yuriy-universal-ivanov** [comment]
@stewartmiles ,
Sure. Done.

**Comment by stewartmiles** [comment]
Thank you.
